.github/workflows/build_and_test.yaml

```diff
@@ -0,0 +1,78 @@
+name: Build and Test PR
+
+on:
+  pull_request: # Trigger for pull requests.
+    types: [opened, synchronize, reopened]
+  workflow_dispatch: # Allows for manual triggering.
+    inputs:
+      ref:
+        description: "The ref to build and test."
+        required: False
+
+jobs:
+  lint:
+    name: Lint
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+        with:
+          path: src
+          ref: ${{ github.event.inputs.ref || github.ref }}
+          # This makes the merge base available for the C++ linter, so that it
+          # can tell which files have changed.
+
+      - name: Lint
+        uses: ./src/.github/workflows/custom-actions/lint-player
+
+  build_and_test:
+    # Don't waste time doing a full matrix of test runs when there was an
+    # obvious linter error.
+    needs: lint
+    strategy:
+      matrix:
+        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
+        browser: ["ChromeHeadless", "FirefoxHeadless", "EdgeHeadless"]
+        exclude:
+          - os: ubuntu-latest
+            browser: EdgeHeadless
+          - os: macos-latest
+            browser: EdgeHeadless
+        include:
+          - os: ubuntu-latest
+            extra_flags: "--use-xvfb"
+          - os: macos-latest
+            # The integration tests don't work so well in macOS, especially for
+            # Firefox (where they crash the browser), so disable them.
+            extra_flags: "--quick"
```

Comment:
I think we can start with this, but we need to address this issue if at all possible.